builtin/providers/aws/resource_aws_emr.go

```diff
@@ -0,0 +1,439 @@
+	conn := meta.(*AWSClient).emrconn
+
+	req := &emr.TerminateJobFlowsInput{
+		JobFlowIds: []*string{
+			aws.String(d.Id()),
+		},
+	}
+
+	_, err := conn.TerminateJobFlows(req)
+	if err != nil {
+		log.Printf("[ERROR], %s", err)
+		return err
+	}
+
+	d.SetId("")
+	return nil
+}
+
+func expandApplications(apps []interface{}) []*emr.Application {
+	appOut := make([]*emr.Application, 0, len(apps))
+
+	for _, appName := range expandStringList(apps) {
+		app := &emr.Application{
+			Name: appName,
+		}
+		appOut = append(appOut, app)
+	}
+	return appOut
+}
+
+func loadGroups(d *schema.ResourceData, meta interface{}) ([]*emr.InstanceGroup, error) {
+	emrconn := meta.(*AWSClient).emrconn
+	reqGrps := &emr.ListInstanceGroupsInput{
+		ClusterId: aws.String(d.Id()),
+	}
+
+	respGrps, errGrps := emrconn.ListInstanceGroups(reqGrps)
+	if errGrps != nil {
+		return nil, fmt.Errorf("Error reading EMR cluster: %s", errGrps)
+	}
+	return respGrps.InstanceGroups, nil
+}
+
+func findGroup(grps []*emr.InstanceGroup, typ string) *emr.InstanceGroup {
+	for _, grp := range grps {
+		if *grp.InstanceGroupType == typ {
+			return grp
+		}
+	}
+	return nil
+}
+
+func expandTags(m map[string]interface{}) []*emr.Tag {
+	var result []*emr.Tag
+	for k, v := range m {
+		result = append(result, &emr.Tag{
+			Key:   aws.String(k),
+			Value: aws.String(v.(string)),
+		})
+	}
+
+	return result
+}
+
+func expandBootstrapActions(bootstrapActions []interface{}) []*emr.BootstrapActionConfig {
+	actionsOut := []*emr.BootstrapActionConfig{}
+
+	for _, raw := range bootstrapActions {
+		actionAttributes := raw.(map[string]interface{})
+		actionName := actionAttributes["name"].(string)
+		actionPath := actionAttributes["path"].(string)
+		actionArgs := actionAttributes["args"].(*schema.Set).List()
+
+		action := &emr.BootstrapActionConfig{
+			Name: aws.String(actionName),
+			ScriptBootstrapAction: &emr.ScriptBootstrapActionConfig{
+				Path: aws.String(actionPath),
+				Args: expandStringList(actionArgs),
+			},
+		}
+		actionsOut = append(actionsOut, action)
+	}
+
+	return actionsOut
+}
+
+func expandConfigures(input string) []*emr.Configuration {
+	configsOut := []*emr.Configuration{}
+	if strings.HasPrefix(input, "http") {
+		readHttpJson(input, &configsOut)
+	} else if strings.HasSuffix(input, ".json") {
+		readLocalJson(input, &configsOut)
+	} else {
+		readBodyJson(input, &configsOut)
+	}
+	log.Printf("[DEBUG] Configures %v\n", configsOut)
+
+	return configsOut
+}
+
+func readHttpJson(url string, target interface{}) error {
+	r, err := http.Get(url)
+	if err != nil {
+		return err
+	}
+	defer r.Body.Close()
+
+	return json.NewDecoder(r.Body).Decode(target)
+}
+
+func readLocalJson(localFile string, target interface{}) error {
+	file, e := ioutil.ReadFile(localFile)
+	if e != nil {
+		fmt.Printf("File error: %v\n", e)
+		log.Printf("[ERROR] %s", e)
+		return e
+	}
+	log.Printf("[DEBUG] %s\n", string(file))
```

Comment:
another debug (need to do a sweep of these, some are relevant, others seem just for development)